File: leet_GRAPH_1743_classic_toposort___old.cc
class Solution {
public:
    vector<int> restoreArray(vector<vector<int>>& adjacentPairs)
    {
        unordered_map<int, vector<int>> D;
        unordered_map<int, int> idg;
        for (vector<int> & v: adjacentPairs)
        {
            int a = v[0], b = v[1];
            D[a].push_back(b);
            D[b].push_back(a);
            idg[a]++;
            idg[b]++;
        }
        int key;
        for (auto it: idg)
        {
            if (it.second == 1)
                key = it.first;
        }
        // cout << key << '\n';
        vector<int> res;
        int len = (int) adjacentPairs.size() + 1;
        set<int>    seen;
        int i = -1;
        while (++i < len)
        {
            seen.insert(key);
            res.push_back(key);
            for (int n: D[key])
            {
                idg[n]--;
                if (!seen.contains(n) && idg[n] < 2)
                {
                    key = n;
                    break ;
                }
            }
        }
        return res;
    }
};

File: leet_GRAPH_1743_classic_toposort___old.py
class Solution:
    def restoreArray(self, adjacentPairs: List[List[int]]) -> List[int]:
        D = defaultdict(list)
        idg = defaultdict(int)
        for a, b in adjacentPairs:
            D[a].append(b)
            D[b].append(a)
            idg[b] += 1
        size = len(adjacentPairs) + 1
        res = []
        seen = set()
        # look for the beginning
        curr = -1e9
        for k, v in idg.items():# enumerate(idg):
            if v == 1:
                curr = k
                break
        # print(D, idg, node)
        for i in range(size):
            seen.add(curr)
            res.append(curr)
            idg[a] += 1
            for n in D[curr]:
                idg[n] -= 1
                if n not in seen and idg[n] < 2:
                    curr = n
                    break
        return res


